src/toolkits/drawing_classifier/drawing_classifier.cpp

```diff
@@ -572,23 +575,200 @@ void drawing_classifier::train(gl_sframe data, std::string target_column_name,
+        std::copy(output_itr, output_itr + num_classes, preds.begin());
+        output_itr += num_classes;
+        writer.write({preds}, 0);
+      }
+    }
+  };
+
+  while (data->has_next_batch()) {
+    // Wait until we have just one asynchronous batch outstanding. The work
+    // below should be concurrent with the neural net inference for that batch.
+    pop_until_size(1);
+
+    result result_batch;
+    result_batch.data_info = data->next_batch(batch_size);
+
+    // Send the inputs to the model.
+    std::map<std::string, shared_float_array> results =
+        backend->predict({{"input", result_batch.data_info.drawings}});
+
+    // Copy the (float) outputs to our (double) buffer and add to the SArray.
+    result_batch.data_info.predictions = results.at("output");
+
+    pending_batches.push(std::move(result_batch));
+  }
+
+  pop_until_size(0);
+
+  return writer.close();
+}
+
 gl_sarray drawing_classifier::predict(gl_sframe data, std::string output_type) {
-  /* TODO: Add code to predict! */
-  return gl_sarray();
+  // by default, it should be "probability" if the value is
+  // passed in through python client
+  if (output_type != "probability_vector" && output_type != "class") {
+    log_and_throw(output_type + " is not a valid option for output_type.  " +
+                  "Expected one of: probability, rank");
+  }
+
+  auto data_itr =
+      create_iterator(data, /* is_train */ false, /* class labels */ {});
+
+  gl_sframe predictions = perform_inference(data_itr.get());
+
+  gl_sarray result = predictions["preds"];
+  if (output_type == "class") {
+    flex_list class_labels = read_state<flex_list>("classes");
+    auto max_prob_label = [=](const flexible_type& ft) {
+      const flex_vec& prob_vec = ft.get<flex_vec>();
+      auto max_it = std::max_element(prob_vec.begin(), prob_vec.end());
+      return class_labels[max_it - prob_vec.begin()];
+    };
+
+    result = result.apply(max_prob_label, class_labels.front().get_type());
+  }
+
+  return result;
 }
 
 gl_sframe drawing_classifier::predict_topk(gl_sframe data,
                                            std::string output_type, size_t k) {
-  /* TODO: Add code to predict_topk! */
-  return gl_sframe();
+  // check valid input arguments
+  if (output_type != "probability" && output_type != "rank") {
+    log_and_throw(output_type +
+                  " is not a valid option for output_type.  "
+                  "Expected one of: probability, rank");
+  }
+
+  // data inference
+  std::unique_ptr<data_iterator> data_it =
+      create_iterator(data, /* is_train */ false, /* class lables */ {});
+
+  gl_sframe dc_predictions = perform_inference(data_it.get());
+
+  // argsort probability to get the index (rank) for top k class
+  // if k is greater than the class number, set it to be class number
+  flex_list class_labels = read_state<flex_list>("classes");
+  k = std::min(k, class_labels.size());
+
+  auto argsort_prob = [=](const flexible_type& ft) {
```

Comment:
Is this lambda actually capturing anything? Why `[=]`?

Reply:
right. I can remove it.

Reply:
the variable `k` is captured.